src/graphics.rs

```diff
@@ -623,20 +611,40 @@ impl Context {
             }
         }
 
-        if self.cache.blend != self.pipelines[pipeline.0].params.color_blend {
-            self.set_blend(self.pipelines[pipeline.0].params.color_blend);
+        self.set_cull_face(self.pipelines[pipeline.0].params.cull_face);
+        self.set_blend(self.pipelines[pipeline.0].params.color_blend);
+        self.set_stencil(self.pipelines[pipeline.0].params.stencil_test);
+        self.set_color_write(self.pipelines[pipeline.0].params.color_write);
+    }
+
+    pub fn set_cull_face(&mut self, cull_face: CullFace) {
+        if self.cache.cull_face == cull_face {
+            return;
         }
 
-        if self.cache.stencil != self.pipelines[pipeline.0].params.stencil_test {
-            self.set_stencil(self.pipelines[pipeline.0].params.stencil_test);
+        match cull_face {
+            CullFace::Nothing => unsafe {
+                glDisable(GL_CULL_FACE);
+            },
+            CullFace::Front => unsafe {
+                glEnable(GL_CULL_FACE);
+                glCullFace(GL_FRONT);
+            },
+            CullFace::Back => unsafe {
+                glEnable(GL_CULL_FACE);
+                glCullFace(GL_BACK);
+            },
         }
+        self.cache.cull_face = cull_face;
+    }
 
-        let pipeline = &self.pipelines[pipeline.0];
-        if self.cache.color_write != pipeline.params.color_write {
-            let (r, g, b, a) = pipeline.params.color_write;
-            unsafe { glColorMask(r as _, g as _, b as _, a as _) }
-            self.cache.color_write = pipeline.params.color_write;
+    pub fn set_color_write(&mut self, color_write: ColorMask) {
```

Comment:
I am a little bit concerned about making those methods `pub`. I am not really against it right now, but I am not sure that this will survive other backends, where pipelines are more real than in opengl. 

Also, they break the idea that render state is completely describable in declarative way as renderpass/pipeline/binding, wich is nice.

Reply:
Yes, I have similar concerns too. We should balance between "pub" and ability to set the pipeline easily. I try to avoid the pub as much as possible. 

Not sure what's the best way to do it yet - perhaps when we have Metal support it will be clearer if it's better to change these pub back to private.

I'm not very familiar with how other frameworks handle similar situations (although reading about it more these days.) If you have suggestions, please mention! I also looked at intel's fastuidraw library that has an "uber shader" and advertises small amount of draw calls.

**Do you know of anything that's using sokol and drawing non-convex polygons with holes? I can look to see how they use sokol there.**


In the concave polygon draw code currently I have something like:
```
set_pipeline(...)
... 
draw_polygon:
    update stencil()
    draw()
    update_cull_face()
    draw()
    disable_stencil()
    draw()
```
I'll look into making it simpler/less draw calls.

Reply:
I made a quick look on metal docs and apparently it is possible to keep those methods public even with metal. 
In sokol/oryol IIRC you will end up with 
```
draw_polygon:
  set_pipeline(stencil_pipeline)
  draw()
  set_piepline(..)
  draw()
```

Same with wgpu, gfx and I guess most vulkan-based APIs. Which is sad for that particular use case, so while we could keep this public - why not?

Reply:
Sounds good to keep public.